### PR TITLE
bitinformation(kwargs)

### PR DESCRIPTION
**Pull request by aaronspring**
- `get_bitinformation` now understands `kwargs`:
  - `confidence`
  - `masked_value` now defaults to `NaN` capturing `np.NaN`s from the input
  - `set_zero_insignificant`
  - `mask` is not allowed and `masked_value` recommended instead

### CODE REVIEW

**Comment by aaronspring** [comment]
Help to fix https://colab.research.google.com/github/ageron/julia_notebooks/blob/master/Julia_for_Pythonistas.ipynb#scrollTo=jKpjTHMbao6r

**Comment by aaronspring** [comment]
Fix in get n plot bitinformation.jl

**Review by observingClouds — CHANGES_REQUESTED** [review]
I like the kwargs approach. However, we should ensure that the Julia function accepts also a list of `dims`. This would help to implement our idea  to retrieve the maximum keepbits. The  functionality itself should however likely be in `get_keepbits`.

**Comment by aaronspring** [comment]
> list of `dims`

I would rather implement `dim` as list in python

**Comment by aaronspring** [comment]
the wrapper has to match booth these functions and use `signed_exponent` if not `Int`:
```julia
function bitinformation(A::AbstractArray{T};
                        dim::Int=1,
                        masked_value::Union{T,Nothing}=nothing,
                        kwargs...) where {T<:Union{Integer,AbstractFloat}}

function bitinformation(A::AbstractArray{T},
                        mask::BitArray;
                        dim::Int=1,
                        set_zero_insignificant::Bool=true,
                        confidence::Real=0.99) where {T<:Union{Integer,AbstractFloat}}
```

**Comment by observingClouds** [comment]
Yes, the part with the signed_exponent is already handled by these two intermediate functions: [1](https://github.com/observingClouds/bitinformation_pipeline/blob/e8b74d43fef8d034556c99f2e2769af4637d0f1e/bitinformation_pipeline/get_n_plot_bitinformation.jl#L4) and [2](https://github.com/observingClouds/bitinformation_pipeline/blob/e8b74d43fef8d034556c99f2e2769af4637d0f1e/bitinformation_pipeline/get_n_plot_bitinformation.jl#L14).

These need to be adjusted to forward the kwargs.

**Comment by aaronspring** [comment]
found a hacky way: passing kwargs as strings in `jl.eval` and preprocess before

File: bitinformation_pipeline/bitinformation_pipeline.py [hunk 1]
```diff
@@ -49,10 +65,21 @@ def get_bitinformation(ds, dim=None, mask=None, label=None, overwrite=False):
...
             X = ds[var].values
             Main.X = X
-            info_per_bit[var] = jl.eval("get_bitinformation(X)")
+            if "mask" in kwargs:
+                raise ValueError(
+                    "bitinformation_pipeline does not wrap the mask argument. Mask your xr.Dataset with NaNs instead."
+                )
+            if "dim" in kwargs:
+                if isinstance(kwargs["dim"], str):
+                    kwargs["dim"] = ds[var].get_axis_num(kwargs["dim"]) + 1
```
Comment (observingClouds) [inline bitinformation_pipeline/bitinformation_pipeline.py:76]: So here we could add

```suggestion
                if isinstance(kwargs["dim"], str):
                      kwargs["dim"] = ds[var].get_axis_num(kwargs["dim"]) + 1
                elif isinstance(kwargs["dim"], list):
                      kwargs["dim"] = [ds[var].get_axis_num(dim) + 1 for dim in kwargs["dim"]]
```

File: bitinformation_pipeline/bitinformation_pipeline.py [hunk 2]
```diff
@@ -49,10 +65,21 @@ def get_bitinformation(ds, dim=None, mask=None, label=None, overwrite=False):
...
+            if "masked_value" not in kwargs:
+                kwargs["masked_value"] = "NaN"
+            kwargs_str = " ," + ", ".join([f"{k}={v}" for k, v in kwargs.items()])
+            kwargs_str = kwargs_str.replace("True", "true").replace("False", "false")
+            logging.debug(f"get_bitinformation(X{kwargs_str})")
+            info_per_bit[var] = jl.eval(f"get_bitinformation(X{kwargs_str})")
```
Comment (observingClouds) [inline bitinformation_pipeline/bitinformation_pipeline.py:82]: And here we could loop over `kwargs["dim"]`. Is that how you think about list support for `dim`?

File: bitinformation_pipeline/bitinformation_pipeline.py (hunk 1 above)
Comment (aaronspring) [reply bitinformation_pipeline/bitinformation_pipeline.py:76]: but does `bitinformation.jl` understand `dim` as `list`? I would implement a loop over dim not here, rather somewhere else

File: bitinformation_pipeline/bitinformation_pipeline.py (hunk 2 above)
Comment (aaronspring) [reply bitinformation_pipeline/bitinformation_pipeline.py:82]: I would implement it at the beginning of `get_bitinformation`:
```python
def get_bitinformation(ds, dim):
    if isinstance(dim, list):
        for d in dim:
            get_bitinformation(ds, d)
        # somehow aggregate d results
```
but lets do this in a different PR. this is getting to large already

**Comment by aaronspring** [comment]
by accident I kick your new code out. I try to merge `main`

**Comment by aaronspring** [comment]
helped to fix the broken test file: https://stackoverflow.com/questions/31233777/python-source-code-string-cannot-contain-null-bytes

**Review by observingClouds — COMMENTED** [review]
This looks good! I'll open a new issue for the support of `dim` as a list. Was it on purpose that the notebook contains output and cell numbers now?

**Comment by aaronspring** [comment]
oh no. I will get rid of the notebook containing output

**Pull request merged**